src/libcore/alloc.rs

```diff
@@ -713,8 +690,8 @@ pub unsafe trait AllocRef {
     /// This function is unsafe because undefined behavior can result
     /// if the caller does not ensure all of the following:
     ///
-    /// * `ptr` must denote a block of memory currently allocated via
-    ///   this allocator,
+    /// * If `layout` is non-zero in size, `ptr` must denote a block of
```

Comment:
I don't think this is strong enough. If implemented for an allocator that does actually allocate for zero-sized allocations (valid, if odd implementation), then trying to free an arbitrary pointer to a zero-sized layout would cause UB.

Rather, this should read something along the lines of "pointer was allocated by this allocator and has not yet been decallocated," which only allows zero-sized layout pointers which were produced by `alloc`.

Of course, implementations can provide stronger guarantees.